Improving update_llc_test_checks.py script

nikolaperic opened the pull request:
Script that automatically generates CHECK directives has been improved to generate outlined function names with arbitrary suffix for Mips target. This means that, inside ```MachineOutliner``` class, the label names of the outlined functions can change their suffix after the ```OUTLINED_FUNCTION``` part of the label and then the ```CHECK``` directives generated by this script will still be correct. In this way, the script is made more flexible.

This also fixes test ```mips_generated_funcs.test```.

Commit:
update_llc_test_checks.py script improvement

Script that generates CHECK directives has been improved to generate
outlined function names with arbitrary suffix for Mips target.
This also fixes test mips_generated_funcs.test.

djtodoro commented:
@nikolaperic should be created against `nanomips` branch.

farazs-github reviewed (changes requested):
Contains a lot of duplicate change from PR #10

djtodoro commented:
> Contains a lot of duplicate change from PR #10

Yes, but the PR was not merged with the `nanomips`, which is the main dev branch as of now.

cme commented on llvm/utils/UpdateTestChecks/asm.py:
```diff
@@ -411,6 +411,9 @@ def get_run_handler(triple):
       handler = s
       best_prefix = prefix
 
+  global target
```
It feels very odd to pass this as a global, and it doesn't seem too closely related; would it be better to check the triple at the use point, perhaps?

nikolaperic replied:
This ```target``` variable is used to check if we can assume that outlined function names start with ```"OUTLINED_FUNCTION"``` string. But since ```"OUTLINED_FUNCTION"``` string comes form ```MachineOutliner``` class, which is not architecture-specific, there is no need to check the target at all. I will remove this.

cme commented on llvm/utils/UpdateTestChecks/common.py:
```diff
@@ -685,7 +691,17 @@ def add_checks(output_lines, comment_marker, prefix_list, func_dict, func_name,
+        # If we are on Mips target check if we generate new name for outlined function that
+        # contains FileCheck variable in it and if so, use that name in CHECK-LABEL directive.
+        is_outlined_fn = func_name.find("OUTLINED_FUNCTION")
+        if is_outlined_fn != -1 and asm.target == "mips":
```
Is this really specific to mips? Or does it apply to anything that has the machine outliner enabled? I don't see anything architecture-specific?

It it is architecture-specific, does it apply to general MIPS as well as nanomips?

nikolaperic replied:
I think you are right - ```"OUTLINED_FUNCTION"``` label name comes from ```MachineOutliner``` class which is not architecture-specific. Inside ```MachineOutliner::createOutlinedFunction()``` nanomips target gets a different suffix for label names of outlined functions but names still start with ```"OUTLINED_FUNCTION"``` string. So I think it should be safe to remove target check.

cme replied:
Great :)

Commit:
Remove target check from update-llc-test-checks.py

Remove target check since outlined function names
start with "OUTLINED_FUNCTION" string regardless
of the target.

cme reviewed (commented):
Looks good, but I'm wondering how did you test it? It's not something that's easy enough to test on its own, I guess you've kept some old `.ll` files to run it over and checked the new results?

nikolaperic commented:
> Looks good, but I'm wondering how did you test it? It's not something that's easy enough to test on its own, I guess you've kept some old `.ll` files to run it over and checked the new results?

The functioning of the script is tested by the ```mips_generated_funcs.test``` test. I relied on that test. The test consists of copying the IR function for which the code outlining will happen, executing the script and checking whether the ```CHECK``` directives are generated as expected. Expected outputs are defined in files ```mips_generated_funcs.ll.generated.expected``` and ```mips_generated_funcs.ll.nogenerated.expected```. Since I also changed the contents of those two files I made sure their new content represent a valid test that does not fail by running the ```llvm-lit``` tool on them. 

When I made sure that the ```mips_generated_funcs.test``` test did not fail, I've changed the way label names of outlined functions were generated inside ```MachineOutliner``` class. More precisely, I changed the suffix of the outlined function names, checked again whether expected outputs are not failing and ran ```mips_generated_funcs.test``` test again.

This is the procedure I used to verify that the script correctly generates ```CHECK``` directives. Do you think this is good enough?

nikolaperic commented:
@cme I guess that this can be merged?

cme commented:
> @cme I guess that this can be merged?

Yep. Thanks! :)
